bpf/include/linux/bpf.h

```diff
@@ -7314,4 +7314,16 @@ struct bpf_iter_num {
 	__u64 __opaque[1];
 } __attribute__((aligned(8)));
 
+
+/* opaque definition of the internal struct bpf_map structure found in
+ * linux/bpf.h
+ *
+ * this structure enables the usage of the bpf_map_for_each_elem and its 
+ * associated callback function. 
+ *
+ * any functions using this structure in Cilium's datapath should never access
+ * a field in the pointer to this structure.
+ */
+struct bpf_map;
```

Comment:
Is this strictly necessary? Instead, can the calling void use a void* instead, like the `void *map` specified in the helper definition? Given this type is never fully declared, I'm also not sure what the `any functions using this structure in Cilium's datapath should never access a field in the pointer to this structure` comment hints at. The compiler doesn't know of any fields, so it can't really emit field accesses.

Reply:
Ah, I think you're right here just using void* will work.
I meant to give this a bit more semantics to not confuse the reading of the eBPF helper (basically it just looks prettier at the call sites with this fwd declaration). But I didn't consider just using a void pointer. 

The comment was really supposed to mean "if you write a function in our datapath with a `struct bpf_map` argument don't plan on using it" as more of a hint to the human planning their own code. 

Happy to change this to void* if we think this is a better or cleaner way to go?

Reply:
(As a C programmer in a previous life, I always preferred using the real types rather than `void *` or typedefs, as those tended to obscure the real meaning of variables)

Reply:
> I always preferred using the real types rather than void * or typedefs

Naturally, but this is supposed to be an opaque pointer/cookie, the caller is not supposed to interact with it. And the type is left unimplemented, so most of the value/clarity will be gained by using a good variable name at the call site. The type name doesn't really matter here.

Reply:
hmm, fair point. Since the value can't be used, maybe void* is **actually** the correct semantic to use. It carries the idea of "opaque value" intrinsically.

Reply:
Okay this was updated to (void *) and the struct was moved from the header.